File: src/render/engine.ts
/**
 * Minimal template rendering engine.
 * Supports: {{var}}, {{nested.var}}, {{#if cond}}...{{/if}}, {{#unless cond}}...{{/unless}}, {{#each items}}...{{/each}}
 */

export interface TemplateContext {
	[key: string]: unknown;
}

function resolve(path: string, ctx: TemplateContext): unknown {
	return path.split('.').reduce<unknown>((obj, key) => {
		if (obj != null && typeof obj === 'object') {
			return (obj as Record<string, unknown>)[key];
		}
		return undefined;
	}, ctx);
}

function isTruthy(val: unknown): boolean {
	if (Array.isArray(val)) return val.length > 0;
	return Boolean(val);
}

export function render(template: string, ctx: TemplateContext): string {
	let result = template;

	// Process {{#each items}}...{{/each}} blocks
	// Consume leading/trailing whitespace when tags are on their own line
	result = result.replace(
		/[ \t]*\{\{#each\s+(\S+?)\}\}\n?([\s\S]*?)[ \t]*\{\{\/each\}\}\n?/g,
		(_match, key: string, body: string) => {
			const items = resolve(key, ctx);
			if (!Array.isArray(items)) return '';
			return items
				.map((item, index) => {
					const itemCtx: TemplateContext = {
						...ctx,
						'.': item,
						'@index': index,
						'@first': index === 0,
						'@last': index === items.length - 1,
					};
					if (typeof item === 'object' && item !== null) {
						Object.assign(itemCtx, item);
					} else {
						itemCtx['this'] = item;
					}
					return render(body, itemCtx);
				})
				.join('');
		},
	);

	// Process {{#if cond}}...{{else}}...{{/if}} blocks
	result = result.replace(
		/[ \t]*\{\{#if\s+(\S+?)\}\}\n?([\s\S]*?)[ \t]*\{\{\/if\}\}\n?/g,
		(_match, key: string, body: string) => {
			const parts = body.split(/[ \t]*\{\{else\}\}\n?/);
			const val = resolve(key, ctx);
			if (isTruthy(val)) {
				return render(parts[0], ctx);
			}
			return parts[1] ? render(parts[1], ctx) : '';
		},
	);

	// Process {{#unless cond}}...{{/unless}} blocks
	result = result.replace(
		/[ \t]*\{\{#unless\s+(\S+?)\}\}\n?([\s\S]*?)[ \t]*\{\{\/unless\}\}\n?/g,
		(_match, key: string, body: string) => {
			const val = resolve(key, ctx);
			if (!isTruthy(val)) {
				return render(body, ctx);
			}
			return '';
		},
	);

	// Process {{variable}} substitutions
	result = result.replace(/\{\{(\S+?)\}\}/g, (_match, key: string) => {
		const val = resolve(key, ctx);
		if (val === undefined || val === null) return '';
		return String(val);
	});

	return result;
}

File: internal/render/engine.go
package render

import (
	"fmt"
	"regexp"
	"strings"
)

// Ctx is a template context (nested map).
type Ctx map[string]any

// Render processes a Handlebars-compatible template with the given context.
// Supports: {{var}}, {{nested.var}}, {{#if}}, {{else}}, {{/if}}, {{#unless}}, {{#each}}.
func Render(template string, ctx Ctx) string {
	result := template

	// Process {{#each items}}...{{/each}}
	eachRe := regexp.MustCompile(`(?s)[ \t]*\{\{#each\s+(\S+?)\}\}\n?(.*?)[ \t]*\{\{/each\}\}\n?`)
	result = eachRe.ReplaceAllStringFunc(result, func(match string) string {
		parts := eachRe.FindStringSubmatch(match)
		key, body := parts[1], parts[2]
		items := resolve(key, ctx)
		arr, ok := toSlice(items)
		if !ok {
			return ""
		}
		var sb strings.Builder
		for i, item := range arr {
			itemCtx := Ctx{}
			// Copy parent context
			for k, v := range ctx {
				itemCtx[k] = v
			}
			itemCtx["."] = item
			itemCtx["@index"] = i
			itemCtx["@first"] = i == 0
			itemCtx["@last"] = i == len(arr)-1
			if m, ok := item.(map[string]any); ok {
				for k, v := range m {
					itemCtx[k] = v
				}
			} else {
				itemCtx["this"] = item
			}
			sb.WriteString(Render(body, itemCtx))
		}
		return sb.String()
	})

	// Process {{#if cond}}...{{else}}...{{/if}}
	ifRe := regexp.MustCompile(`(?s)[ \t]*\{\{#if\s+(\S+?)\}\}\n?(.*?)[ \t]*\{\{/if\}\}\n?`)
	result = ifRe.ReplaceAllStringFunc(result, func(match string) string {
		parts := ifRe.FindStringSubmatch(match)
		key, body := parts[1], parts[2]
		elseBlocks := regexp.MustCompile(`(?s)[ \t]*\{\{else\}\}\n?`).Split(body, 2)
		val := resolve(key, ctx)
		if isTruthy(val) {
			return Render(elseBlocks[0], ctx)
		}
		if len(elseBlocks) > 1 {
			return Render(elseBlocks[1], ctx)
		}
		return ""
	})

	// Process {{#unless cond}}...{{/unless}}
	unlessRe := regexp.MustCompile(`(?s)[ \t]*\{\{#unless\s+(\S+?)\}\}\n?(.*?)[ \t]*\{\{/unless\}\}\n?`)
	result = unlessRe.ReplaceAllStringFunc(result, func(match string) string {
		parts := unlessRe.FindStringSubmatch(match)
		key, body := parts[1], parts[2]
		val := resolve(key, ctx)
		if !isTruthy(val) {
			return Render(body, ctx)
		}
		return ""
	})

	// Process {{variable}} substitutions
	varRe := regexp.MustCompile(`\{\{(\S+?)\}\}`)
	result = varRe.ReplaceAllStringFunc(result, func(match string) string {
		parts := varRe.FindStringSubmatch(match)
		key := parts[1]
		val := resolve(key, ctx)
		if val == nil {
			return ""
		}
		return fmt.Sprintf("%v", val)
	})

	return result
}

func resolve(path string, ctx Ctx) any {
	keys := strings.Split(path, ".")
	var current any = map[string]any(ctx)
	for _, key := range keys {
		m, ok := current.(map[string]any)
		if !ok {
			return nil
		}
		current = m[key]
	}
	return current
}

func isTruthy(val any) bool {
	if val == nil {
		return false
	}
	switch v := val.(type) {
	case bool:
		return v
	case string:
		return v != ""
	case int:
		return v != 0
	case float64:
		return v != 0
	case []any:
		return len(v) > 0
	case []string:
		return len(v) > 0
	}
	return true
}

func toSlice(v any) ([]any, bool) {
	if v == nil {
		return nil, false
	}
	switch s := v.(type) {
	case []any:
		return s, true
	case []string:
		result := make([]any, len(s))
		for i, item := range s {
			result[i] = item
		}
		return result, true
	}
	return nil, false
}
